paper.tex

```diff
@@ -449,14 +455,83 @@ \subsection{Pipeline Design} \label{subsec:pipeline-design}
+        t1 $\gets$ t0 + dt\;
+        (next\_p, next\_q) $\gets$ eval\_cubic\_and\_deriv(cubic, t1)\;
+        \tcp{Estimate error of geometric}
+        \tcp{Hermite interpolation to Euler spiral}
+        \nosemic(th0, th1, chord\_len, err) $\gets$ fit\_cubic\_to\_es(\;
+        \pushline\dosemic last\_p, next\_p, last\_q, next\_q, dt)\;
+        \popline \eIf{err <= TOL}{
+            es $\gets$ euler\_segment(th0, th1, last\_p, next\_p)\;
+            \tcp{Find location of any cusp in outline.}
+            \tcp{If $0 \leq t < 1$, draw evolute at 't'.}
+            cusp0 $\gets$ es\_curvature(es, 0) * offset + chord\_len\;
+            cusp1 $\gets$ es\_curvature(es, 1) * offset + chord\_len\;
+            t $\gets$ cusp0 * cusp1 >= 0 ? 1 : cusp0 / (cusp0 - cusp1)\;
+            \If{cusp0 >= 0}{
+                evolute\_finalize()\;
+                es\_seg\_flatten\_offset(es, offset, vec2(0, t))\;
+            }
+            \If{cusp0 < 0 OR t < 1}{
+                evolute\_start(es, offset)\;
+                range $\gets$ cusp0 >= 0 ? vec2(t, 1) : vec2(0, t)\;
+                es\_seg\_flatten\_evolute(es, range)\;
+                es\_seg\_flatten\_offset\_reverse(es, offset, range)\;
+            }
+            \If{cusp0 < 0 AND t < 1}{
+                evolute\_finalize()\;
+                es\_seg\_flatten\_offset(es, offset, vec2(t, 1))\;
+            }
+            \tcp{Pop the stack}
+            t0\_u $\gets$ t0\_u + 1u\;
+            shift $\gets$ countTrailingZeros(t0\_u)\;
+            t0\_u $\gets$ t0\_u >> shift;
+            dt $\gets$ dt * f32(1u << shift)\;
+            last\_p $\gets$ next\_p\;
+            last\_q $\gets$ next\_q\;
+        }{
+            \tcp{Error is above the tolerance.}
+            \tcp{Push the stack; Continue subdivision}
+            t0\_u $\gets$ t0\_u * 2u\;
+            dt $\gets$ dt * 0.5\;
+        }
+    }
+}
+\caption{Outline of the iterative cubic Bézier flattening routine invoked in our compute shader. A working WGSL implementation is provided in the supplemental materials.}
+\label{alg:flatten-cubic}
+\end{algorithm}
+
+All input segments first get converted to a cubic Bézier by degree elevation. Our flattening routine (see Algorithm~\ref{alg:flatten-cubic}) iteratively subdivides the input cubic segment in half, and at each step, computes an error estimate for geometric Hermite interpolation to an Euler spiral segment. Once the error satisfies a desired tolerance threshold, the Euler spiral segment is directly flattened to a polyline (or, alternately, converted to a sequence of arc segments) by invoking a subroutine that is parameterized by the stroke halfwidth (see \texttt{es\_seg\_flatten\_offset} in Algoritm~\ref{alg:flatten-cubic}). For a regular fill, this parameter is set to $0$. For strokes, the sign of the parameter determines the position of the parallel curve relative to the source segment as well as the winding direction of the output lines.
```

Comment:
"iteratively subdivides...in half" suggests it always does at least one such subdivision. I'd add "when the error tolerance is exceeded" or "whenever".

Reply:
PTAL at the new phrasing.